Source/JavaScriptCore/builtins/PromiseConstructor.js

```diff
@@ -78,6 +76,109 @@ function all(iterable)
     return promiseCapability.@promise;
 }
 
+@linkTimeConstant
+function promiseOnRejectedWithContext(argument, context)
+{
+    "use strict";
+
+    @rejectPromiseWithFirstResolvingFunctionCallCheck(context.globalContext.promise, argument);
+}
+
+@linkTimeConstant
+function promiseAllOnFulfilled(argument, context)
+{
+    "use strict";
+
+    var globalContext = context.globalContext;
+    var values = globalContext.values;
+
+    @putByValDirect(values, context.index, argument);
+
+    if (!--globalContext.remainingElementsCount)
+        @resolvePromiseWithFirstResolvingFunctionCallCheck(globalContext.promise, values);
+}
+
+@linkTimeConstant
+function promiseNewOnRejected(promise)
+{
+    "use strict";
+
+    return function @reject(reason) {
+        return @rejectPromiseWithFirstResolvingFunctionCallCheck(promise, reason);
+    };
+}
+
+@linkTimeConstant
+function promiseAllNewResolveElement(globalContext, index)
+{
+    "use strict";
+
+    var alreadyCalled = false;
+    return (argument) => {
+        if (alreadyCalled)
+            return @undefined;
+        alreadyCalled = true;
+
+        var values = globalContext.values;
+        @putByValDirect(values, index, argument);
+
+        if (!--globalContext.remainingElementsCount)
+            return @resolvePromiseWithFirstResolvingFunctionCallCheck(globalContext.promise, values);
+    };
+}
+
+function all(iterable)
```

Comment:
is a similar optimization relevant for `InternalPromise`? Could be good for ESM loader perf

Reply:
Because `Promise.all` can have many user-observable behavior, we should not use it for internal promise, let's use `@InternalPromise.internalAll` instead :)